### PR TITLE
Environments

### PR DESCRIPTION
**Pull request by yuval-hazaz**
- [x] New model Environment
- [x] New model Deployment
- [x] Environment Module (Service, Resolver, Test)
- [x] Create sandbox environment when creating a new app

### CODE REVIEW

File: packages/amplication-server/src/core/environment/environment.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    }).compile();
+
+    service = module.get<EnvironmentService>(EnvironmentService);
+  });
+
+  it('should be defined', () => {
+    expect(service).toBeDefined();
+  });
+
+  it('should create an app role', async () => {
```
Comment (iddan) [inline packages/amplication-server/src/core/environment/environment.service.spec.ts:59]: Update test description

File: packages/amplication-server/src/core/environment/environment.service.spec.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        address: EXAMPLE_ENVIRONMENT_ADDRESS,
+        app: { connect: { id: EXAMPLE_APP_ID } }
+      }
+    };
+    expect(await service.createEnvironment(args)).toEqual(EXAMPLE_ENVIRONMENT);
+    expect(prismaEnvironmentCreateMock).toBeCalledTimes(1);
+    expect(prismaEnvironmentCreateMock).toBeCalledWith(args);
+  });
+
+  it('should find one app role', async () => {
```
Comment (iddan) [inline packages/amplication-server/src/core/environment/environment.service.spec.ts:73]: Update test description

File: packages/amplication-server/src/core/environment/environment.service.spec.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    const args = { where: { id: EXAMPLE_ENVIRONMENT_ID } };
+    expect(await service.getEnvironment(args)).toEqual(EXAMPLE_ENVIRONMENT);
+    expect(prismaEnvironmentFindOneMock).toBeCalledTimes(1);
+    expect(prismaEnvironmentFindOneMock).toBeCalledWith(args);
+  });
+
+  it('should update an app role', async () => {
```
Comment (iddan) [inline packages/amplication-server/src/core/environment/environment.service.spec.ts:80]: Update test description

File: packages/amplication-server/src/core/environment/environment.service.ts [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  CreateEnvironmentArgs,
+  UpdateOneEnvironmentArgs
+} from './dto';
+import { FindOneArgs } from 'src/dto';
+
+@Injectable()
+export class EnvironmentService {
+  constructor(private readonly prisma: PrismaService) {}
+
+  async createEnvironment(args: CreateEnvironmentArgs): Promise<Environment> {
```
Comment (iddan) [inline packages/amplication-server/src/core/environment/environment.service.ts:15]: Why is this method not exposed by the resolver?

File: packages/amplication-server/src/core/environment/environment.service.ts [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    return this.prisma.environment.create(args);
+  }
+
+  async getEnvironment(args: FindOneArgs): Promise<Environment | null> {
```
Comment (iddan) [inline packages/amplication-server/src/core/environment/environment.service.ts:19]: Why is this method not exposed by the resolver?

File: packages/amplication-server/prisma/migrations/20201006094009/README.md [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
++  steps       ActionStep[]
++  builds      Build[]
++  deployments Deployment[]
+ }
+ model ActionStep {
+   id          String           @id @default(cuid())
+@@ -361,6 +364,47 @@
+   action         Action          @relation(fields: [actionId], references: [id])
+   actionId       String
++  deployments Deployment[]
```
Comment (iddan) [inline packages/amplication-server/prisma/migrations/20201006094009/README.md:97]: Do we have to have this field?

File: packages/amplication-server/prisma/migrations/20201006094009/README.md [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
++  build         Build                @relation(fields: [buildId], references: [id])
++  buildId       String
++  environment   Environment          @relation(fields: [environmentId], references: [id])
++  environmentId String
++  status        EnumDeploymentStatus
++  message       String?
++  action        Action               @relation(fields: [actionId], references: [id])
++  actionId      String
++
++
```
Comment (iddan) [inline packages/amplication-server/prisma/migrations/20201006094009/README.md:138]: Unneeded extra space

File: packages/amplication-server/src/core/app/app.service.ts [hunk 8]
```diff
@@ -51,6 +54,12 @@ export class AppService {
         },
         roles: {
           create: USER_APP_ROLE
+        },
+        environments: {
+          create: {
+            name: DEFAULT_ENVIRONMENT_NAME,
+            address: cuid()
```
Comment (iddan) [inline packages/amplication-server/src/core/app/app.service.ts:61]: Please source from an environment variable or add a todo comment.

**Review by iddan — COMMENTED** [review]
Good job! Please see my comments

File: packages/amplication-server/src/core/environment/environment.service.spec.ts (hunk 1 above)
Comment (yuval-hazaz) [reply packages/amplication-server/src/core/environment/environment.service.spec.ts:59]: Already done in the following PR

File: packages/amplication-server/src/core/environment/environment.service.spec.ts (hunk 2 above)
Comment (yuval-hazaz) [reply packages/amplication-server/src/core/environment/environment.service.spec.ts:73]: Already done in the following PR

File: packages/amplication-server/src/core/environment/environment.service.spec.ts (hunk 3 above)
Comment (yuval-hazaz) [reply packages/amplication-server/src/core/environment/environment.service.spec.ts:80]: Already done in the following PR

File: packages/amplication-server/src/core/environment/environment.service.ts (hunk 4 above)
Comment (yuval-hazaz) [reply packages/amplication-server/src/core/environment/environment.service.ts:15]: Currently, each application only have a single environment which is created by default with the application.
At this stage, the user cannot add more environments.

File: packages/amplication-server/src/core/environment/environment.service.ts (hunk 5 above)
Comment (yuval-hazaz) [reply packages/amplication-server/src/core/environment/environment.service.ts:19]: The current flow requires minimal interaction with the environment. We only need to know its ID in order to send it when calling "CreateDeployment".
I already added in the next PR a FieldResolver on app to get the app environments in order to get the single env ID.

Later down the road, we will implement the rest of the functionality which is required for multi env.

File: packages/amplication-server/prisma/migrations/20201006094009/README.md (hunk 6 above)
Comment (yuval-hazaz) [reply packages/amplication-server/prisma/migrations/20201006094009/README.md:97]: I have no idea why this field appears here... in the actual schema.prisma file, the ActionStep model does not have this field

File: packages/amplication-server/prisma/migrations/20201006094009/README.md (hunk 7 above)
Comment (yuval-hazaz) [reply packages/amplication-server/prisma/migrations/20201006094009/README.md:138]: Did you notice that you are review the auto-generated .md file ?

File: packages/amplication-server/src/core/app/app.service.ts (hunk 8 above)
Comment (yuval-hazaz) [reply packages/amplication-server/src/core/app/app.service.ts:61]: The address only represents the subdomain and it must be globally unique. By default the address is based on a CUID but the users have an option to choose a custom sub domain later.
As you move forward with the rest of the deployment code, if you see a reason to include the full domain name in the address, we can come back to this code.

**Comment by yuval-hazaz** [comment]
> Good job! Please see my comments

Thanks. I replied to all comments with no code changes. Please review again.

File: packages/amplication-server/prisma/migrations/20201006094009/README.md (hunk 7 above)
Comment (iddan) [reply packages/amplication-server/prisma/migrations/20201006094009/README.md:138]: No sorry

File: packages/amplication-server/src/core/app/app.service.ts (hunk 8 above)
Comment (iddan) [reply packages/amplication-server/src/core/app/app.service.ts:61]: Oh... I see. Okay, we'll come back to this later.

**Review by iddan — APPROVED** [review]

**Pull request merged**